Code/src/main/java/nl/utwente/group10/haskell/expr/Apply.java

```diff
@@ -1,66 +1,60 @@
 package nl.utwente.group10.haskell.expr;
 
+import nl.utwente.group10.haskell.env.Env;
 import nl.utwente.group10.haskell.exceptions.HaskellTypeError;
+import nl.utwente.group10.haskell.hindley.GenSet;
+import nl.utwente.group10.haskell.hindley.HindleyMilner;
 import nl.utwente.group10.haskell.type.FuncT;
 import nl.utwente.group10.haskell.type.Type;
 
-import java.util.Arrays;
-
 /**
- * A type of expression that is a lazy application of a function with certain arguments.
+ * Lazy application of an argument to a function.
```

Comment:
Hmm, but analyze() doesn't evaluate the expression either, does it? It seems the laziness you describe here seems to be a bit different from Haskell's special kind of laziness, I think that's part of the confusion.

I'd just call it an application.